File: is_prime.cpp
#include <iostream>
using namespace std;

/*
template<bool B, class T = void>
struct enable_if {};

template<class T>
struct enable_if<true, T> { typedef T type; };
 */

template <int p, int i>
struct check_prime {
    static const bool value = (p % i != 0) && (check_prime<p, i - 1>::value);
};

template <int p>
struct check_prime<p, 1> {
    static const bool value = true;
};

template <int p>
struct is_prime {
    static const bool value = check_prime<p, p - 1>::value;
};

template <int p>
struct ModuloInt {
    int value;

    ModuloInt(int v): value(p) {}
    ModuloInt<p> operator+ (ModuloInt<p> o) {
        return ModuloInt((o.value + value) % p);
    }

    template <typename = typename std::enable_if<is_prime<p>::value >::type >
    ModuloInt<p> invert() {
        return 0 /*dummy*/;
    }
};

int main() {
    cout << is_prime<12>::value << endl;
    ModuloInt<12> v = 13;
    v = v + 9;
    v.invert();
}

File: is_prime.nim
import macros, strutils

type
  ModuloInt[p: static[int]] = distinct int

proc isPrime(v: int): bool =
  for j in 2..<v:
    if v mod j == 0:
      return false

  return true

# f(a, b, c)
# a.f(b, c)
# f1(f2(f3(x))) x.f3.f2.f1

proc invert[p: static[int]](m: ModuloInt[p]): ModuloInt[p] =
  static:
    if not isPrime(p):
      error("attempting to invert a number modulo not-prime ($1)!" % [$p])
  return ModuloInt[p](0) # dummy

let v = ModuloInt[12](5)
discard v.invert()
